Realm/RLMAsyncTask.mm

```diff
@@ -474,3 +475,82 @@ - (void)wait:(void (^)())completion {
     completion();
 }
 @end
+
+@implementation RLMAsyncSubscriptionTask {
+    RLMUnfairMutex _mutex;
+
+    RLMSyncSubscriptionSet *_subscriptionSet;
+    dispatch_queue_t _queue;
+    NSTimeInterval _timeout;
+    void (^_completion)(NSError *);
+
+    bool _called;
+    dispatch_block_t _worker;
+}
+
+- (instancetype)initWithSubscriptionSet:(RLMSyncSubscriptionSet *)subscriptionSet
+                                  queue:(nullable dispatch_queue_t)queue
+                                timeout:(NSTimeInterval)timeout
+                             completion:(void(^)(NSError *))completion {
+    if (!(self = [super init])) {
+        return self;
+    }
+
+    _subscriptionSet = subscriptionSet;
+    _queue = queue;
+    _timeout = timeout;
+    _completion = completion;
+
+    return self;
+}
+
+- (void)waitForSubscription {
+    std::lock_guard lock(_mutex);
+
+    if (_timeout != 0) {
+        // Setup timer
+        dispatch_time_t time = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(_timeout * NSEC_PER_SEC));
+        // If the call below doesn't return after `time` seconds, the internal completion is called with an error.
+        _worker = dispatch_block_create(DISPATCH_BLOCK_ASSIGN_CURRENT, ^{
+            NSString* errorMessage = [NSString stringWithFormat:@"Waiting for update timed out after %.01f seconds.", _timeout];
+            NSError* error = [NSError errorWithDomain:NSPOSIXErrorDomain code:ETIMEDOUT userInfo:@{NSLocalizedDescriptionKey: errorMessage}];
+            [self invokeCompletionWithError:error];
+        });
+
+        dispatch_after(time, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), _worker);
+    }
+
+    [self waitForSync];
+}
+
+-(void)waitForSync {
+    if (_completion) {
+        _subscriptionSet->_subscriptionSet->get_state_change_notification(realm::sync::SubscriptionSet::State::Complete)
+            .get_async([self](realm::StatusWith<realm::sync::SubscriptionSet::State> state) noexcept {
+                NSError *error = makeError(state);
+                [self invokeCompletionWithError:error];
+            });
+    }
+}
+
+-(void)invokeCompletionWithError:(NSError * _Nullable)error {
+    std::lock_guard lock(_mutex);
+
+    if (_called) {
+        return;
+    }
+    _called = true;
+
+    if (_worker) {
+        dispatch_block_cancel(_worker);
+        _worker = nil;
+    }
+
+    if (_queue) {
+        return dispatch_async(_queue, ^{
+            _completion(error);
+        });
+    }
+    _completion(error);
```

Comment:
I cannot remove the `called` flag, this is the only way to avoid a second call when the `getAsync` is called after the timeout callback, this can happen if the connection is restored or the subscription is taking too much time.

Reply:
You can just check if `completion` is non-nil, as it should always be nil once it's been called (or when another thread is about to call it), and if there's ever a case where it isn't then that's a bug.

Reply:
updated this to not use `_called`